### PR TITLE
chore: re-enable generic attribute tests

### PR DESCRIPTION
**Pull request by jammehcow**
This PR aims to re-enable the generic attribute tests that were commented out in efe74e52015bafee049086f3f526378be673ed98. This appears to have been removed due to compilation errors when using the generic attributes feature of C# 11.

I've split this work out from #321 / #319 as that PR should be purely about fixing the derived attribute classes issue.

TODO:
 - [x] Upgrade test projects to .NET 7 (required as .NET 6 projects will use the .NET standard Vogen package which will not have the generic attributes)
 - [x] Re-enable `GenericAttributeTests`
 - [x] Validate that tests still produce expected results
 - [x] (maybe) fill in any untested gaps

### CODE REVIEW

**Comment by jammehcow** [comment]
Bumping the test projects from .NET 6 to .NET 7 has caused the `System.Runtime` assembly to jump from `6.0.0.0` to `7.0.0.0` which breaks all other tests now 😅

Since `System.Runtime` isn't just a NuGet package you can add (not anymore) this might require something a bit funky to get working - I'm thinking getting the test projects to target net6.0 & net7.0 then only tests targeting 7.0 in the net7.0 assembly. I don't like it but I'm not seeing any alternatives with the current setup.

**Comment by jammehcow** [comment]
I've introduced the `XUnit.SkippableFact` NuGet package to allow tests to be skipped when they're not going to do anything - specifically in the case of these tests when they won't verify on .NET 6. There will be first-party support for skipping tests in xunit 3 whenever that comes out ([see here](https://github.com/xunit/xunit/issues/2073)).

**Comment by jammehcow** [comment]
As for the state of this PR, I've done the following:
 - Explicitly set the SDK version for the entire solution to .NET 7 (Rider was defaulting to 6 and getting confused)
 - Re-enable the generic attribute tests
   - Fixed one of the tests which was non-compilable (cf4b304c019b346269681dfdd72a72bc31d3f8e7)
 - Changed the targeting of all test projects from .NET 6 to multi-target both .NET 6 & .NET 7
   - This effectively doubles the test count, but I have added some skips into the snapshot tests for single-framework tests
 - Added filtering in the snapshot & test runner to prevent tests targeting .NET 7 from running under .NET 6 - and vice versa
 - Fixed one of the `TODO`s around testing the analzyer with a generic attribute as the attribute from the Vogen package can be used with .NET 7
   - Added another test skip in here for targets below .NET 7 as they're useless to test

File: global.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (SteveDunn) [inline global.json:1]: Do we still need this?

**Comment by SteveDunn** [comment]
@jammehcow - great stuff - lovely to see those tests finally running, well done! 👍

**Review by SteveDunn — APPROVED** [review]
LGTM - thank you 👍

File: global.json (hunk 1 above)
Comment (jammehcow) [reply global.json:1]: Don't *need* it, but it's a nice to have when using .NET CLI commands. Happy to remove if you don't want it 👍

**Pull request merged**